src/Init/Lean/PrettyPrinter/Parenthesizer.lean

```diff
@@ -0,0 +1,444 @@
+/-
+Copyright (c) 2020 Sebastian Ullrich. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sebastian Ullrich
+-/
+
+/-!
+The parenthesizer inserts parentheses into a `Syntax` object where syntactically necessary, usually as an intermediary
+step between the delaborator and the formatter. While the delaborator outputs structurally well-formed syntax trees that
+can be re-elaborated without post-processing, this tree structure is lost in the formatter and thus needs to be
+preserved by proper insertion of parentheses.
+
+# The abstract problem & solution
+
+The Lean 4 grammar is unstructured and extensible with arbitrary new parsers, so in general it is undecidable whether
+parentheses are necessary or even allowed at any point in the syntax tree. Parentheses for different categories, e.g.
+terms and levels, might not even have the same structure. In this module, we focus on the correct parenthesization of
+parsers defined via `Lean.Parser.prattParser`, which includes both aforementioned built-in categories. Custom
+parenthesizers can be added for new node kinds, but the data collected in the implementation below might not be
+appropriate for other parenthesization strategies.
+
+Usages of a parser defined via `prattParser` in general have the form `p rbp`, where `rbp` is the right-binding power.
+Recall that a Pratt parser greedily parses a leading token with precedence at least `rbp` (otherwise it fails) followed
+by zero or more trailing tokens with precedence *higher* than `rbp`. Thus we should parenthesize a syntax node `stx`
+produced by `p rbp` if
+
+1. the left-most token in `stx` has precedence < `rbp`, or
+2. the left-most token to *the right of* `stx` has precedence > `rbp` (because without parentheses, `p rbp` would have
+   parsed it as well and made it a part of `stx`).
+
+Note that in case 2, it is also sufficient to parenthesize a *parent* node as long as the offending token is still to
+the right of that node. For example, imagine the tree structure of `(f $ fun x => x) y` without parentheses. We need to
+insert *some* parentheses between `x` and `y` since the lambda body is parse with precedence 0, while `y` as an
+identifier has precedence `appPrec`. But we need to parenthesize the `$` node anyway since the precedence of its
+RHS (0) again is smaller than that of `y`. So it's better to only parenthesize the outer node than ending up with
+`(f $ (fun x => x)) y`.
```

Comment:
Not sure I interpreted this remark correctly. I am assuming you are making a point for introducing parentheses as a top-down operation. If we do it top-down, then we get `(f $ fun x => x) y`. If we do it bottom-up, we would get `(f $ (fun x => x)) y`. Is that right?

Reply:
Roughly, yes. It's a bit complicated by the fact that the necessary information (the precedence of the lambda body) is propagated bottom-up. Since we do everything in a single pass, it's perhaps more accurate to say we *delay* the decision bottom-up until the last possible moment (the furthest parent node), instead of calling it top-down.